From b0087f0173a47c942696c01dd1ade96aa8d42e6a Mon Sep 17 00:00:00 2001
From: Dan Nica <dan.nica@nxp.com>
Date: Sun, 21 Feb 2021 12:09:27 +0200
Subject: [PATCH] s32g: mmc: Fix while loops and remove unneeded udelay()

The while loops are written incorrectly, as they allow the execution to
move forward if INT_STATUS is 0, resulting in reading CMD_RSPn registers
before they are updated by the controller. Now with proper checking of
the INT_STATUS register, the udelay() call is no longer needed.

Issue: ALB-6480
Signed-off-by: Dan Nica <dan.nica@nxp.com>
---
 drivers/nxp/s32g/mmc/s32g_mmc.c | 24 ++++++++----------------
 1 file changed, 8 insertions(+), 16 deletions(-)

diff --git a/drivers/nxp/s32g/mmc/s32g_mmc.c b/drivers/nxp/s32g/mmc/s32g_mmc.c
index 11ee52386..4904d006c 100644
--- a/drivers/nxp/s32g/mmc/s32g_mmc.c
+++ b/drivers/nxp/s32g/mmc/s32g_mmc.c
@@ -1,5 +1,5 @@
 /*
- * Copyright 2020 NXP
+ * Copyright 2020-2021 NXP
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -257,17 +257,11 @@ static int s32g274a_mmc_send_cmd(struct mmc_cmd *cmd)
 	mmio_write_32(USDHC_CMDARG, cmd->cmd_arg);
 	mmio_write_32(USDHC_CMD_XFR_TYP, cmd_xfr_typ);
 
-	/* Without a delay here, the values read from USDHC_CMD_RSP(x)
-	 * may not be up-to-date.
-	 */
-	udelay(1000);
-
-	while ((regdata = mmio_read_32(USDHC_INT_STATUS))) {
+	do {
+		regdata = mmio_read_32(USDHC_INT_STATUS);
 		if (regdata & INT_STATUS_CMD_ERROR)
 			goto cmd_error;
-		if (regdata & INT_STATUS_CC)
-			break;
-	}
+	} while (!(regdata & INT_STATUS_CC));
 
 	if (cmd->resp_type & MMC_RSP_136) {
 		for (i = 0; i < 4; i++)
@@ -280,14 +274,12 @@ static int s32g274a_mmc_send_cmd(struct mmc_cmd *cmd)
 		cmd->resp_data[0] = mmio_read_32(USDHC_CMD_RSP(0));
 	}
 
-	if (adtc_mask & (BIT(cmd->cmd_idx))) {
-		while ((regdata = mmio_read_32(USDHC_INT_STATUS))) {
+	if (adtc_mask & (BIT(cmd->cmd_idx)))
+		do {
+			regdata = mmio_read_32(USDHC_INT_STATUS);
 			if (regdata & (INT_STATUS_DATA_ERROR))
 				goto data_error;
-			if (regdata & (INT_STATUS_TC | INT_STATUS_DINT))
-				break;
-		}
-	}
+		} while (!(regdata & (INT_STATUS_TC | INT_STATUS_DINT)));
 
 	return 0;
 
-- 
2.17.1

